### DISCUSSION TITLE
The "new" mympd.service file should not be default

### DISCUSSION THREAD

**Discussion by GK-774** [jcorporation/myMPD#920] [General]
First of all "systemd-analyze security" is not a security assesment tool, it only looks at the sandbox features built into systemd. It does not check the service itself. So it is safe to ignore these if you have a valid security concept for your system and a security concept needs to take in account the whole system, possible attack vectors, etc.; it´s not an easy task, just "switching on" some sandboxing in systemd gives a false sense of security.

Anyway, the new mympd.service file has several problems in my view:

- It has DynamicUser=yes but then has User=mympd and Group=mympd specified. From the docs: "Note that if User= is specified and the static group with the name exists, then it is required that the static user with the name already exists. Similarly, if Group= is specified and the static user with the name exists, then it is required that the static group with the name already exists."

- myMPD maintains files in /var/lib/mympd and /var/cache/mympd so it should have a proper static user anyway (like basically all other services do as well) who owns these.

- DynamicUser aready implies ProtectSystem=strict, ProtectHome=read-only, PrivateTmp=yes and RemoveIPC=yes
but these are redundantly specified as well.

- Availability of these options depend on the systemd version, e.g. on an older RedHat EL7 the script does not work at all, on EL8 it starts but myMPD does not appear to work correctly with all features. (I did not bother to debug that in detail as I see no need to use any of these options).

As it is, it's not generic, does not work "out of the box" in many cases and its usefullness is debateable (and untested). It is apparently targeted at one specific Debian distribution and therefore should be marked as such, optional for those who are willing to experiment with it, and not the default. It should not be enforced but be a deliberate choice and integration on a specific distribution should be done by the package maintainer. (Personally I use RedHat Enterprise compatible distributions, they have selinux on per default and can confirm that the service files RedHat ships sometimes have ProtectHome=yes and PrivateTmp=true but thats it.)

**Comment by jcorporation** [comment]
Thanks for opening this discussion.
We should involve @dvzrv and @MichaIng

**Comment by MichaIng** [comment]
Of course sandboxing a service doesn't mean the whole server security is guaranteed, even only regarding the service itself. That neither developer nor admins can rely on this protecting against everything, is, or should be, clear to everyone, but is no argument to just not do sandboxing. It is an additional layer of security around the particular service, which should generally be considered, if there are no downsides. Mainly it may protect against, in whichever way, injected harmful executables, 3rd party libraries and, of course, security and privacy holes in the software itself.

Compared to a static service user, a dynamic user has the benefit that it's impossible to grant it any additional permissions to access non-world accessible files, via group or chown. And the dirs it requires access to (state and cache dirs), are chowned automatically by systemd.

> "Note that if User= is specified and the static group with the name exists, then it is required that the static user with the name already exists. Similarly, if Group= is specified and the static user with the name exists, then it is required that the static group with the name already exists."

Right, but the package postinst script does already not create the user and group anymore, but now actively removes them, if systemd is used. For 3rd party/other distro packages, it would need to be done accordingly, of course, but then one can also add an own systemd service or alter it, if really required.

> myMPD maintains files in /var/lib/mympd and /var/cache/mympd so it should have a proper static user anyway (like basically all other services do as well) who owns these.

As said, both dirs are setup by systemd automatically on service start, so that the dynamic user has access. Outside of the service scope, only root has access:
```sh
ls -al /var/{lib,cache}/private
```

> DynamicUser aready implies ProtectSystem=strict, ProtectHome=read-only, PrivateTmp=yes and RemoveIPC=yes
but these are redundantly specified as well.

That is true, those can be removed.

> Availability of these options depend on the systemd version

The latest relevant settings are StateDirectory/CacheDirectory, added with systemd v235, released 2017-10-06, part of Ubuntu Bionic and Debian Buster, which I personally would consider sufficient legacy OS support. Does latest myMPD, respectively the builds from the official repositories, work at all on older versions, e.g. RHEL7 🤔?

> on EL8 it starts but myMPD does not appear to work correctly with all features

Sounds like something to investigate separately. If the service starts, then it is likely a setup issue. Or which particular features, which do not work correctly, do you mean? I could imagine that some distros setup MPD a way that the socket is not world-writable, probably this is true on RHEL8? This, like generally access to non-world-accessible files/dirs can be achieved with the `SupplementaryGroups=` systemd option.

**Comment by jcorporation** [comment]
> mympd user/group

Only the debian postinst script removes the user/group. This is not implemented for rpms.

> on EL8 it starts but myMPD does not appear to work correctly with all features

Is it the scripting feature with a script that calls commands with sudo? This is rectricted by the hardend systemd unit: https://github.com/jcorporation/myMPD/discussions/916

> RHEL 7

This peace of software is far too old, to be explicitly supported by myMPD.

> DynamicUser aready implies ProtectSystem=strict, ProtectHome=read-only, PrivateTmp=yes and RemoveIPC=yes
but these are redundantly specified as well

The unit file defines `ProtectSystem=full`. Removing unnecessary options is preferable, but eventually @dvzrv added this options for a good reason.

**Reply by MichaIng** [reply]
> Only the debian postinst script removes the user/group. This is not implemented for rpms.

I see. But the user/group also is not created by the FPM package, or is this somehow implied? https://github.com/jcorporation/myMPD/blob/master/contrib/packaging/rpm/mympd.spec

`ProtectSystem=strict` is stricter than `ProtectSystem=full`, and while I haven't tested, I'm sure `ProtectSystem=full` is overridden in this case:
- https://www.freedesktop.org/software/systemd/man/systemd.exec.html#ProtectSystem=
- https://www.freedesktop.org/software/systemd/man/systemd.exec.html#DynamicUser=

I guess (no offence) dvzrv simply didn't know this, I didn't know either before reading it in another context recently.

**Reply by jcorporation** [reply]
Thats means removing `ProtectSystem` is safe, I tested it.

Removing `ProtectHome` leads on my ubuntu kinetic to a lower score for `systemd-analyze security`.

**Reply by MichaIng** [reply]
Yes, and for `User` and `Group` I also do not see any reasons. In fact, removing those two should also solve any issues if the user and/or group exists. But not sure whether myMPD internally requires to be run by `mympd` user or otherwise tries to fork off?

**Reply by jcorporation** [reply]
myMPD tries only to drop privileges to the mympd user if it is started as root.

**Reply by MichaIng** [reply]
Okay, then it should be fine. I can actually do some testing the next minutes.

**Reply by jcorporation** [reply]
Great. The devel branch includes some fixes for the postinst script: https://github.com/jcorporation/myMPD/issues/917

**Reply by MichaIng** [reply]
Btw, to check whether a service is actually running:
```sh
! systemctl -q is-active mympd || systemctl stop mympd
```

Tested with the four settings removed, and watching to music here 🙂:
```console
root@VM-Bullseye:~# systemctl cat mympd
# /lib/systemd/system/mympd.service
#
# SPDX-License-Identifier: GPL-3.0-or-later
# myMPD (c) 2018-2022 Juergen Mang <mail@jcgames.de>
# https://github.com/jcorporation/mympd
#
[Unit]
Description=myMPD server daemon
Documentation=man:mympd(1)
Requires=network.target local-fs.target
After=mpd.service

[Service]
AmbientCapabilities=CAP_NET_BIND_SERVICE
CapabilityBoundingSet=CAP_NET_BIND_SERVICE
DynamicUser=yes
ExecStart=/usr/bin/mympd
LockPersonality=yes
MemoryDenyWriteExecute=yes
PrivateDevices=yes
ProtectClock=yes
ProtectControlGroups=yes
ProtectHostname=yes
ProtectKernelLogs=yes
ProtectKernelModules=yes
ProtectKernelTunables=yes
ProtectProc=invisible
RestrictRealtime=yes
StateDirectory=mympd
CacheDirectory=mympd
RestrictAddressFamilies=AF_INET AF_INET6 AF_NETLINK AF_UNIX
RestrictNamespaces=yes
SystemCallArchitectures=native
SystemCallFilter=@system-service

[Install]
WantedBy=multi-user.target
```

**Reply by jcorporation** [reply]
great and thanks for the hint!

**Comment by GK-774** [comment]
Well ok, this is absolutely not what I meant to say. What I meant to say it should not be default because it does not work out of the box the way it is now on every system where mpd and myMPD otherwise work without problems. (And yes, although I use something more recent, it does work perfectly on EL7 which is supported until June 30, 2024, BTW). RedHat uses the sandboxing features of systemd very sparsely and likely for good reason, it has selinux enabled per default with well thought out (and tested!) policies. These sandboxing features are a solution in search of a problem here, this is not an Android phone.

Additionally I certainly do not want to engage in a discussion about how to implement system security as this usually leads to nothing fruitful, it´s to a degree distribution specific and I´m aware that those exited by the latest and greatest features of systemd usually do not listen to those that have been doing this for a living almost three decades. 

So my points are:
- myMPD should continue to work out of the box on any reasonably distribution. (not only on one specific Debian release).
- Those who want to use systemd sandboxing features for some reason, by all means go for it, if you know what you are doing (If you break it you keep all the pieces).
- Implementing distribution specific configurations should be done by those making a package for this distribution.

As long as such changes do not interfere with the usability on other systems I don´t care, but I´m concerned it is steering in this way.

**Reply by MichaIng** [reply]
> What I meant to say it should not be default because it does not work out of the box the way it is now on every system where mpd and myMPD otherwise work without problems.

Can you be more specific what exactly does not work? As said, all I tested, works pretty well OOTB. I even recognised that myMPD itself does not require even read access to music and playlist directories, this is all done by MPD only. So zero filesystem access is required outside of state and cache directory.

> These sandboxing features are a solution in search of a problem here, this is not an Android phone.

IMO this is the wrong approach to deal with security features (from software developer and admin POV). The other way it makes sense: Use every sandboxing feature you can, and disable those where you find good reasons to do so. Of course distros by default uses moderate security settings, since they cannot know what admins are gonna do and there it is a balance between security and functionality/setup effort for admins. But as an end user software developer, you know exactly what your software is expected to do and what it requires access and permissions for, and what not, so tight hardening for your particular software/service is easier and makes sense.

**Reply by jcorporation** [reply]
> So zero filesystem access is required outside of state and cache directory.

This is not correct. myMPD needs read access to the music directory for some features (lyrics, enhanced coverimage support).

> > These sandboxing features are a solution in search of a problem here, this is not an Android phone.

> IMO this is the wrong approach to deal with security features

Right, I try to release myMPD with secure defaults.

**Reply by MichaIng** [reply]
> This is not correct. myMPD needs read access to the music directory for some features (lyrics, enhanced coverimage support).

Thanks for clarification. Metadata are generally shown, but lyrics and enhanced cover images are then not handled by MPD, I suppose.

**Reply by GK-774** [reply]
"this is the wrong approach to deal with security features" no, it´s avoiding cargo cult science where untested features make you think you are safe.

"distros by default uses moderate security settings" - no, not all, e.g. RedHat has all relevant certifications for misson critical systems (including government and military use).

"hardening for your particular software" - yes, e.g. proper privilege separation and/or a comprehensive selinux policy would be, but systemd is not part of the software here it´s part of the OS.

**Reply by MichaIng** [reply]
I think there is no point in discussion whether the example systemd service shipped with the repo and its own packaging should make vast use of sandboxing features or not, without knowing what exactly those break in your case 😉. So:

> Can you be more specific what exactly does not work?

**Reply by GK-774** [reply]
You still did not get my point. Im using RedHat and there these features are in general only used sparsely for a few services (mainly those which belong to systemd itself). Debian has disabled selinux per default, RedHat has it enabled per default.
I don´t care which options are in the mympd.service file for Debian and I´m not going to debug it on RedHat because I´m not using it.
What I do care about if changes are made to the software itself to accommodate systemd features. For the time being it does not break anything, it should stay that way.

**Reply by MichaIng** [reply]
So you want to have additional hardening removed from myMPD's systemd service, only because RH, which you do not use, doesn't use them (often)? 🤔

> What I do care about if changes are made to the software itself to accommodate systemd features.

None are done so far, aside of that `mympd -c` allows to proceed also if the `mympd` user does not exist, which however does make sense in general, IMO. Not sure what makes you worry that myMPD is adapted to systemd? It's the other way round, the systemd unit is adapted to what myMPD really needs.

**Reply by jcorporation** [reply]
myMPD shoud be compatible with any sort of init system. I run myMPD with my Alpine Linux based myMPDos image. Alpine Linux uses openrc.

**Reply by GK-774** [reply]
"because RH, which you do not use" just the opposite, I use RedHat (since 1995 and since 2000 professionally). "None are done so far" yes, and thats good. A selinux policy (like mpd has) probably would be useful, but that´s for another day.

**Reply by GK-774** [reply]
> I run myMPD with my Alpine Linux based myMPDos image. Alpine Linux uses openrc.

Looks interesting, I will have a look at Alpine tomorrow (the x86_64 version a rasperry I do not have).

**Reply by jcorporation** [reply]
I like alpine because it is really small and can run on a raspberry completely in the ram.

**Comment by jcorporation** [comment]
> myMPD should continue to work out of the box on any reasonably distribution. (not only on one specific Debian release).

This should be the case. The service unit changes are pushes from an Arch Linux packager ;)

I think only the migration to DynamicUser for distributions with systemd was not good enough tested.

**Reply by GK-774** [reply]
As long I´m not forced to use those changes it´s ok with me.

**Reply by MichaIng** [reply]
Ah sorry, I indeed did not understand you concern 😅. No these hardenings are not required for myMPD, they only limit the executed process to do anything it is not supposed to do, aside of, as mentioned direct read-access to non-world-readable media files, which would then, if you do use the hardenings, require the SupplementaryGroups setting as well to gain access via group permissions.

**Reply by GK-774** [reply]
You do not need to explain me how it works - by the way "hardening" is a bit exaggerated. "these hardenings are not required for myMPD" certainly not on aproperly set up system, let alone one with selinux enabled where most of these are souperflous.

**Reply by GK-774** [reply]
"mympd -c allows to proceed also if the mympd user does not exist" - you think that´s a good idea?

**Reply by jcorporation** [reply]
This is a good idea, because a static mympd user is with systemd optional.

**Reply by GK-774** [reply]
It´s probably not a good idea. A service that maintains files shall have a defined UID and GID and that´s the standard for good reason on every sensible configured system. From https://www.redhat.com/sysadmin/systemd-secure-services  :

"When should a regular static user be used for a service, and when is a dynamic one preferred? Dynamic users are great when the user identity is ephemeral, and no integration with other services in the system is needed. But when we have a policy in the database to allow specific user access, directories shared with a particular group, or any other configuration where we want to refer to the user name, dynamic users are probably not the best option.
...
Returning to the subject of DynamicUser=yes, such transient users are only possible when the service is not allowed to create permanent files on disk. If such files were visible to other users, they would be shown as having no owner, or worse, they could be accessed by the new transient user with the same number, leading to an information leak or an unintended privilege escalation."

**Reply by dvzrv** [reply]
>  certainly not on aproperly set up system, let alone one with selinux enabled where most of these are souperflous.

That is assuming a bit much about the target. Not every system uses selinux.
Judging from your assertions in https://github.com/jcorporation/myMPD/discussions/920#discussion-4629439 it is unclear to me why you would want to downplay the sandboxing options that systemd offers. Is this solely based on the premise, that they do not work "perfectly" on RHEL7? If that's the case you have to get more specific and maybe the packaged files can be adapted for that distribution, or you could add overrides for them.

If anything, this discussion shows, that myMPD by default assumes that it has access to every file on the system (which is a problematic design decision from a security perspective).
Even *if* the service was not using `DynamicUser=yes`, that does not change anything about the fact, that there are only a few ways to achieve access to an arbitrary user's music collection directory: By running as root (the wooden hammer method), by specifically allowing access to a directory to the service user (e.g. using a service override with `ReadWritePaths=` or `ReadOnlyPaths=`, which may involve `SupplementaryGroups=`, etc.) or by having the service user own all music collection files.
As myMPD is a C based network facing service, option two is definitely the preferred one in this use-case IMHO.

**Reply by GK-774** [reply]
"That is assuming a bit much about the target. Not every system uses selinux."
- It´s you who makes assumptions. Not every system uses systemd "sandboxing" features istead.

"hat myMPD by default assumes that it has access to every file on the system" - Nosense. it has never been that way since the dawn of Unix.

"running as root" - Nonsense again. It appears you have not understood what was written above. (myMPD drops privileges correctly BTW, as it should be).
This leads nowhere, good luck with your pseudo security, I´m out here.

**Reply by jcorporation** [reply]
myMPD writes only to his state and cache directories and reading the files from the mpd music and playlist directory is optional. As I understand the enabled sandboxing options correctly, there is no need for a `ReadOnlyPaths=` override. Adding a group membership could be necessary with old or new service unit.

The only thing I am unsure is, whether the `DynamicUser=` adds more complexity as it brings benefits.

Users have problems with this:
#912 #929 #930

**Reply by jcorporation** [reply]
> myMPD drops privileges correctly 

That is correct, myMPD even refuses to run, if it could not drop his privileges.

**Reply by dvzrv** [reply]
@jcorporation in case you want to remove `DynamicUser=yes`, it is of course still possible to only use `User=` and `Group=` in the service file. However, you'd need to ensure the respective user exists then (that's what I wanted to circumvent with `DynamicUser=yes` as myMPD does not use a whole lot of files by itself from what I could tell - and maybe I was wrong to assume this is easier for users).
One way of doing this on systemd based systems is by providing a [sysusers.d](https://man.archlinux.org/man/core/systemd/sysusers.d.5.en) integration. E.g.:
```
u mympd - "myMPD user" -
```

The above file contents would lead to a user and group `mympd` being created (without a `$HOME` and without a fixed UID/GID), if they don't exist yet.

@GK-774 No need for aggression and baseless accusations. If you can't lead a civil discussion here without using a condescending tone towards others, maybe this discussion is indeed not for you.

**Reply by jcorporation** [reply]
I tend to revert back to legacy user creation in the postinst script of the package. I think this is what users expect and this is what we must do, if we install myMPD on a none-systemd distribution. Some consistency across distributions would be good, this reduces the support hassle.

@MichaIng: What do you think about this?

**Reply by dvzrv** [reply]
@jcorporation I'll use sysusers.d integration for the Arch Linux package then, as we do not create users in install scripts anymore.
Thanks for the feedback!

**Reply by jcorporation** [reply]
Interesting, if Suse, Fedora and Debian goes the same way the sysusers.d concept should be preferred.

**Reply by john-** [reply]
> Users have problems with this: #912 #929 #930

To be clear the last 2 issues are for me.  #929 is be with same problem as #912.  #930 is startup error that occurs after I applied fixes to get things working in #912.  I am running on older ARM based Ubuntu (18) so not sure how general my problem is.

I had to set DynamicUser=no as we are hosting a party this weekend and need music!

There is no upgrade option for my Ubuntu SBC.  So instead of installing a new version clean I am going to test out Armbian on it.  Maybe that will address the error with DynamicUser option.

**Reply by MichaIng** [reply]
Ubuntu 18.04 does not yet support the `StateDirectory`/`CacheDirectory` settings, I think. So the two dirs are not chowned to the dynamic user automatically on service start. The systemd journal should then also show two related warnings that these settings are ignored.

Backwards compatibility and consistency across systemd and non-systemd systems are indeed arguments to revert `DynamicUser`. But actually I like the fact that it does not rely on any manually set permissions or created users but does setup everything correctly automatically (on service level, instead of root-privileged process level). The question is how long legacy OS versions shall be supported and whether the dynamic user inconsistency case really adds maintenance burden on supported OS versions. In theory also the systemd version can be checked to edit/override the dynamic user, but that is too probably too much complexity.

The problem with `sysusers.d` is that is is systemd-only. But for package maintainers of distros which do enforce systemd (or packages which do), nice indeed. However, for myMPD's own packages, proper `useradd` has the same result and is universally compatible.

**Reply by jcorporation** [reply]
@MichaIng: Thanks for your opinion.

If this was only an issue with old systemd then its acceptable. I have no distinct policy how long myMPD supports a given distribution. I can add a hint, that the default myMPD service unit works only with systemd v235 and above.

**Reply by john-** [reply]
>  I have no distinct policy how long myMPD supports a given distribution. 

In fact my older version of Ubuntu doesn't have a new enough version of mpd or cmake for myMPD.  I need to compile these from source for myMPD to build/fully function.  Adding a hint for older systemd would be reasonable.

**Reply by MichaIng** [reply]
If other workarounds are required anyway, then I think it is fine to not revert the systemd unit only because of Ubuntu 18.04 an alike. A note somewhere makes sense, just not sure whether to put it, just a comment right into the systemd unit above `DynamicUser=`?

**Reply by john-** [reply]
A documentation update to [Running](https://jcorporation.github.io/myMPD/running) might work.

At the end of **Systemd usage**, something like:

If the version of systemd your distribution uses does not support the `DynamicUser` option, than add static `mympd` user/group and change file permissions accordingly.

**Reply by MichaIng** [reply]
Ah right, seems to fit best there, or additionally. Jürgen is quick: https://github.com/jcorporation/myMPD/commit/5e71c53

**Reply by john-** [reply]
> Ah right, seems to fit best there, or additionally. Jürgen is quick: [5e71c53](https://github.com/jcorporation/myMPD/commit/5e71c53)

Quick and more succinct :)
